### PR TITLE
Sanitize string when printing in JavaGenerator and KotlinGenerator

### LINKED ISSUE #1486: Delimited toString Output
The output of a generated `Message`'s `toString()` can be difficult to read in scenarios with heavily nested messages. I want something like  pretty printing for `toString()`, but without having to worry about the bloat of introducing lots of whitespace into the `toString()` response itself. What I would like to propose is properly delimited `toString()` output so a separate tool can then format that String.

At present `toString()` will return something like this:
```
Person{name=Jadzia Dax, age=1,300, greeting=Hi, I'm {CURRENT_HOST} dax!, file=C:\Users\Dax.bat}
```
Writing a reliable formatter for this will be hard:
- `age=1,` looks like the end of the field
- A template format `{CURRENT_HOST}` could potentially look like a new internal message.

I propose escaping characters `,{}[]\` so we can reliably format the output of `toString()`
The example above would now become:
```
Person{name=Jadzia Dax, age=1\,300, greeting=Hi\, I'm \{CURRENT_HOST\} Dax!, file=C:\\Users\\Dax.bat}
```

### PR DESCRIPTION
**Pull request by oldergod**
fixes #1486

### CODE REVIEW

File: wire-library/wire-runtime/src/commonMain/kotlin/com/squareup/wire/internal/Internal.kt [hunk 1]
```diff
@@ -148,3 +148,13 @@ fun countNonNull(a: Any?, b: Any?, c: Any?, d: Any?, vararg rest: Any?): Int {
   }
   return result
 }
+
+/** Return a string where `,\]\{}` are escaped with a `\`. */
+fun sanitize(value: String): String {
+  return value.replace("""([,\\\[\]{}])""".toRegex(), """\\$1""")
```
[the remark this answers is not in the document]
Comment (JakeWharton) [reply wire-library/wire-runtime/src/commonMain/kotlin/com/squareup/wire/internal/Internal.kt:154]: Cache it. Does not get optimized.

[the remark this answers is not in the document]
Comment (oldergod) [reply wire-library/wire-runtime/src/commonMain/kotlin/com/squareup/wire/internal/Internal.kt]: Extracted into a static field.

File: wire-library/wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt [hunk 2]
```diff
@@ -798,6 +798,7 @@ class KotlinGenerator private constructor(
   }
 
   private fun generateToStringMethod(type: MessageType, nameAllocator: NameAllocator): FunSpec {
+    val sanitizeMember = MemberName("com.squareup.wire.internal", "sanitize")
```
Comment (Egorand) [inline wire-library/wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt:801]: Neat

File: wire-library/wire-runtime/src/commonMain/kotlin/com/squareup/wire/internal/Internal.kt [hunk 3]
```diff
@@ -148,3 +148,15 @@ fun countNonNull(a: Any?, b: Any?, c: Any?, d: Any?, vararg rest: Any?): Int {
   }
   return result
 }
+
+private val ESCAPED_CHARS = """([,\\\[\]{}])""".toRegex()
+
+/** Return a string where `,\]\{}` are escaped with a `\`. */
```
Comment (Egorand) [inline wire-library/wire-runtime/src/commonMain/kotlin/com/squareup/wire/internal/Internal.kt:154]: Escaped chars list in the doc is inconsistent with `ESCAPED_CHARS`.

File: wire-library/wire-runtime/src/commonMain/kotlin/com/squareup/wire/internal/Internal.kt [hunk 4]
```diff
@@ -148,3 +148,15 @@ fun countNonNull(a: Any?, b: Any?, c: Any?, d: Any?, vararg rest: Any?): Int {
...
+fun sanitize(value: String): String {
+  return value.replace(ESCAPED_CHARS, """\\$1""")
+}
+
+/** Return a string where `,\]\{}` are escaped with a `\`. */
```
Comment (Egorand) [inline wire-library/wire-runtime/src/commonMain/kotlin/com/squareup/wire/internal/Internal.kt:159]: Same

File: wire-library/wire-runtime/src/commonMain/kotlin/com/squareup/wire/internal/Internal.kt (hunk 3 above)
Comment (aerb) [reply wire-library/wire-runtime/src/commonMain/kotlin/com/squareup/wire/internal/Internal.kt:154]: Good catch

Comment (oldergod) [reply wire-library/wire-runtime/src/commonMain/kotlin/com/squareup/wire/internal/Internal.kt]: fixed

File: wire-library/wire-runtime/src/commonMain/kotlin/com/squareup/wire/internal/Internal.kt (hunk 4 above)
Comment (oldergod) [reply wire-library/wire-runtime/src/commonMain/kotlin/com/squareup/wire/internal/Internal.kt]: fixed

**Comment by oldergod** [comment]
Replaced the `replace` with a buildString + loop because JS was bugging, see https://youtrack.jetbrains.com/issue/KT-38000

File: wire-library/wire-runtime/src/jvmTest/kotlin/com/squareup/wire/internal/InternalTest.kt [hunk 5]
```diff
@@ -30,4 +30,17 @@ class InternalTest {
     assertEquals(4, countNonNull("xx", "xx", "xx", "xx", null))
     assertEquals(5, countNonNull("xx", "xx", "xx", "xx", "xx"))
   }
+
+  @Test fun sanitizeStrings() {
```
Comment (JakeWharton) [inline wire-library/wire-runtime/src/jvmTest/kotlin/com/squareup/wire/internal/InternalTest.kt:34]: Can we move this whole file to `commonTest`?

File: wire-library/wire-runtime/src/commonMain/kotlin/com/squareup/wire/internal/Internal.kt [hunk 6]
```diff
@@ -148,3 +148,20 @@ fun countNonNull(a: Any?, b: Any?, c: Any?, d: Any?, vararg rest: Any?): Int {
   }
   return result
 }
+
+private const val ESCAPED_CHARS = ",[]{}\\"
+
+/** Return a string where `,[]{}\` are escaped with a `\`. */
+fun sanitize(value: String): String {
+  return buildString {
```
[the remark this answers is not in the document]
Comment (JakeWharton) [reply wire-library/wire-runtime/src/commonMain/kotlin/com/squareup/wire/internal/Internal.kt:156]: Or consider adding a bit extra? Maybe 5? Not sure how common we expect these escapes to be...

File: wire-library/wire-runtime/src/commonMain/kotlin/com/squareup/wire/internal/Internal.kt [hunk 7]
```diff
@@ -148,3 +148,20 @@ fun countNonNull(a: Any?, b: Any?, c: Any?, d: Any?, vararg rest: Any?): Int {
...
+    value.forEach { char ->
+      if (char in ESCAPED_CHARS) append("\\")
+      append(char)
+    }
+  }
+}
+
+/** Return a string where `,[]{}\` are escaped with a `\`. */
+fun sanitize(values: List<String>): String {
+  return values.map(::sanitize).toString()
```
[the remark this answers is not in the document]
Comment (oldergod) [reply wire-library/wire-runtime/src/commonMain/kotlin/com/squareup/wire/internal/Internal.kt:166]: Yeah we were relying on this before so didn't think much of changing it

File: wire-library/wire-runtime/src/commonMain/kotlin/com/squareup/wire/internal/Internal.kt [hunk 8]
```diff
@@ -148,3 +148,20 @@ fun countNonNull(a: Any?, b: Any?, c: Any?, d: Any?, vararg rest: Any?): Int {
...
   return result
 }
+
+private const val ESCAPED_CHARS = ",[]{}\\"
+
+/** Return a string where `,[]{}\` are escaped with a `\`. */
+fun sanitize(value: String): String {
+  return buildString {
+    value.forEach { char ->
+      if (char in ESCAPED_CHARS) append("\\")
```
[the remark this answers is not in the document]
Comment (oldergod) [reply wire-library/wire-runtime/src/commonMain/kotlin/com/squareup/wire/internal/Internal.kt]: fixed

File: wire-library/wire-runtime/src/commonMain/kotlin/com/squareup/wire/internal/Internal.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (oldergod) [reply wire-library/wire-runtime/src/commonMain/kotlin/com/squareup/wire/internal/Internal.kt]: fixed without the extra bit.

File: wire-library/wire-runtime/src/commonMain/kotlin/com/squareup/wire/internal/Internal.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (oldergod) [reply wire-library/wire-runtime/src/commonMain/kotlin/com/squareup/wire/internal/Internal.kt]: fixed

File: wire-library/wire-runtime/src/jvmTest/kotlin/com/squareup/wire/internal/InternalTest.kt [hunk 9]
```diff
@@ -30,4 +30,17 @@ class InternalTest {
...
+    assertEquals("""\,""", sanitize(""","""))
+    assertEquals("""\{""", sanitize("""{"""))
+    assertEquals("""\}""", sanitize("""}"""))
+    assertEquals("""\[""", sanitize("""["""))
+    assertEquals("""\]""", sanitize("""]"""))
+    assertEquals("""\\""", sanitize("""\"""))
+    assertEquals("""Hi\, I'm \{CURRENT_HOST\} dax!""", sanitize("""Hi, I'm {CURRENT_HOST} dax!"""))
+
+    assertEquals(listOf("""\,""", """\{""", """\}""", """\[""", """\]""", """\\""").toString(),
```
[the remark this answers is not in the document]
Comment (oldergod) [reply wire-library/wire-runtime/src/jvmTest/kotlin/com/squareup/wire/internal/InternalTest.kt]: fixed

File: wire-library/wire-runtime/src/jvmTest/kotlin/com/squareup/wire/internal/InternalTest.kt (hunk 5 above)
Comment (oldergod) [reply wire-library/wire-runtime/src/jvmTest/kotlin/com/squareup/wire/internal/InternalTest.kt:19]: I moved the whole class and seems good

**Pull request merged**